base/rational.jl

```diff
@@ -83,6 +83,13 @@ end
 
 function show(io::IO, x::Rational)
     show(io, numerator(x))
+
+    if isone(denominator(x)) && (
+        get(io, :compact, false) || get(io, :typeinfo, Number) <: Rational
+    )
```

Comment:
```suggestion
    if isone(denominator(x)) && get(io, :typeinfo, Any) <: Rational
```
Compact is not enough a substitute for typeinfo <: Rational.

```julia
julia> Union{Int, Float64}[7 3/6; 6/3 2]
2×2 Matrix{Union{Float64, Int64}}:
 7    0.5
 2.0  2

julia> Union{Int, Rational{Int}}[7 3//6; 6//3 2]
2×2 Matrix{Union{Rational{Int64}, Int64}}:
 7  1//2
 2   2
```
Also, if typeinfo is not provided, all we can assume is Any, not Number (though the two are equivalent in this case).